personalrobotics/ada_feeding
# Add Pre MoveTo Idiom

# Description

Before any call to MoveTo, we should do the following: (a) toggle the `watchdog_listener` off; (b) re-tare the F/T sensor; (c) toggle the `watchdog_listener` on; and (d) set force-torque thresholds. The reason for (a) and (c) is that re-taring causes the force-torque sensor to momentarily stop publishing, which kills the watchdog listener.

In order to make it easy to add those calls before any MoveTo node, this PR does the following:

1. Creates an idiom, `pre_moveto_config` that does the above 4 steps.
2. In service of the above, creates an idiom, `retry_call_ros_service`, to make it easier to call ros services and verify the response.
3. Implements the `MoveToConfigurationWithFTThresholdsTree` that adds the `pre_moveto_config` before a `MoveToConfiguration` call.
4. Implements `MoveToRestingPosition`, `MoveAbovePlate`, and `MoveToStowLocation` as `MoveToConfigurationWithFTThresholdsTree`.

# Testing procedure

1. Pull the latest [py_trees_ros (branch: `amaln/service_client`) code](https://github.com/personalrobotics/py_trees_ros/tree/amaln/service_client) and re-run `colcon build`
2. Start the F/T sensor: `ros2 run forque_sensor_hardware forque_sensor_hardware --ros-args -p host:=xxx.xxx.x.xx`
3. Echo the F/T sensor topic: `ros2 topic echo /wireless_ft/ftSensor1`
4. Start the ADA feeding nodes: `ros2 launch ada_feeding ada_feeding_launch.xml`
5. Run MoveIt with the Watchdog Listener: `ros2 launch ada_moveit demo_feeding.launch.py`
6. Verify the following:
    - [x] The F/T sensor readings are not mean 0 (this is common if it hasn't yet been re-tared).
    - [x] Run `ros2 action send_goal /MoveAbovePlate ada_feeding_msgs/action/MoveTo "{}" --feedback`
        - [x] Verify that the MoveIt launchfile prints the messages that indicate that the watchdog listener was turned off, and then turned back on.
        - [x] Verify that the F/T sensor readings become mean-0 (a sign that it re-tared).
        - [x] Verify that the MoveAbovePlate action succeeds.

# Before opening a pull request

# Before Merging
